Adds a movement looping system, replaces inbuild procs and spacedrift with it

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This is gonna be another one of those, alright?
Ok. So like a year ago I realized that the walk() procs are effectively sleep loops 
That means the mc gets a lot less time then it expects, because it's being consumed by a separate scheduler.
This is why everything but the meteors slow down so hard during a meteor storm. They move with a delay of one, and that's a lot of work to do.
This mostly pissed me off because of what it did to atmos, so I tried moving them over to throwing. oranges put me in my place #55749
(Oh by the by, I gave this a shot in march, roh wanted some form of priority system, I burned out, it died. Old me was a coward. #55860)

That brings me to what I'm pring here. I'm replacing the move procs with a new framework. Works like this.
We've got a selection of move_loop datums. Each type holds information about what's going to be moved, and how.
So you've got a move_loop type for moving in a direction, two different flavors for moving towards something, one for moving randomly.
Then we have a movement_packet datum. This tracks what move_loops are attached to an object, and which move loop is currently running.
There's a few rules around move loops. Primarily, you can only have one move loop active at once. This is determined by the loops precedence. Secondly, you're only allowed to have one loop per subsystem.

On that note, the system supports more then one processing subsystem.
The idea is to split up loops into either sets that should overwrite each other, or into groupings of things that should be separate.
The hope here is 1000 space drifters shouldn't fuck over the guy fighting bubblegum on lavaland.

You interact with the system by using one of a few procs that live on SSmove_manager, which exists to act as a namespace so things can use move() as a proc name.

Going back to something I mentioned before, I've integrated spacedrift into this system. So instead of using the old subsystem, and storing like 5 vars on movable, we handle the spacedrift custom logic in a component, and do the moving with just the basic directional mover. I wasn't able to move everything over, but most things made the jump.

I've done a few other things too. Because spacedrifting is no longer randomly slightly delayed before starting because of how processing works, and because we know (at least in theory I've not tested it under high load) how much the subsystem is dying, we can smooth the initial space move.

Additionally, because I'm doing spacedrifting in a component instead of as a move loop, I can track when someone is finishing a drift, and block player movement until they fully land on the turf.

The combination of these two things means starting and stopping a spacedrift is now totally smooth, outside some kinks in the glide size scaling that I may need to work out, and some headache around stopping movement for the last move that I may not be able to. Also some jumping I still haven't squashed

Note: Jetpacks still make it look bad. There's no real way around this without further work done on movement delay that I'm not going to do here. Thanks
I've done my best to capture all this with obs, but my pc is crap. May have to take me at my word to some degree.

https://user-images.githubusercontent.com/58055496/140264402-5233f802-066d-4e61-bde3-d10c7475cc06.mp4

You may have noticed me mentioning something odd in there.
The way movement subsystems work right now is by iterating all currently running loops, no matter how high or low their delay. I'm doing it like this because I'm not sure if a bucketing system would actually be more efficient for the amount of things being moved. May change this in future as the system is expanded.

Lemme see uhhh
I've converted everything that used the walk procs over to my new system, this'll mostly help with large amounts of aggressive simple mobs
Also converted some other things that were using sleep loops, I think just the cursed blob for now. I've got my eyes on effects like sparks, which shouldn't be hard, and things like conveyor belts and disposal holders, which require some backend work. I'd eventually like to tie the system into pressure movement, but that's a bit far away.
Edit:
I've converted sparks over, which means roughly 2 less timers per light, and no more sleeping. Should make mass light breaking slightly less painful
I'm also working on other effects and associates. I've got extinguishers and explosions done, working on some others
This list may well grow, there's a lot of stuff doing replaceable loops, either with timers or for the real old stuff, just sleeps.
Edit Edit:
Holy shit man.
Conveyors, all effects, sprays, most forced movement including slips, ai movement, jps, etc.
I've had to add extra functionality to support some of these.
1: You can pass in an extra "packet" of information that the movement loop keeps track of.
I'm doing this so ai movement can well, work without creating separate datums for each mob, which seemed important. it's a little messy, but I think? it's ok.
2: You can set movement loops apart from the priority system, so they'll just run until you manually stop them, and they won't stop anything else from working. Needed this for conveyors.
Edit: Edit: Edit:
Transit tubes now too

Another barely related change, the move procs no longer coexist/override space drift. I'll need to do a pass on this at some point, but for now poly can fly and bubbles's charge ignores it too.

Notably secborgs do not, eob has assured me this is canonical 
![image](https://user-images.githubusercontent.com/58055496/143566096-be524672-6fef-4f7c-abc7-38f40336735c.png)

Oh and the movement for sprays currently looks like this out out up out out instead of up out out out out. This might change fire extinguishers effectiveness? still kinda torn on if it's worth keeping.

On the topic of performance, I think this is on par with the byond implementation, if not a little faster somehow. I have no idea why.
The operation of adding a loop is a little bit convoluted (especially for space drift, component adds a bit extra on the top), but the looping itself is close to standard. I think this'll be a minor hit for spacedrift, because I'm avoiding some of its optimizations in the name of good feeling space movement, but I don't think that's a huge issue. It comes with the massive bonus of integrating a lot of move types with the mc, which means they can be properly scheduled. 

Ah this is important. I've changed meteor behavior somewhat, along with some other stuff.
Lemme do my best to explain why.

So byond's walk_towards implementation basically looks like this
```
while(true){
    step(moving, get_step(moving, get_dir(moving, target)))
    sleep(delay)
}
```
Trouble is, a step can only be in one of 8 directions.
So if the target is two tiles to the right and 10 tiles up, it'll move up and to the right twice, and then just up the rest of the way.
What I've done is implement a version of walk_towards that effectively gets the rise over run between two objects, and steps at least 1 tile per fire.
So instead of moving two tiles to the right right at the start of the movement, my system would move a tile to the right once every 5 up steps.

This leads to nice lines instead of harsh corners, and I think will make things look a lot nicer.
As a consequence meteors are now much less likely to take the same path as each other, and more likely to go on diagonals. This makes them a bit more deadly.

Also cleaned up meteor code a bit, but that's not a huge deal.
   
Oh right I found a bug in immovable rod code, it didn't properly parent call in part of its moved override, so if you orbited it through a z level transition the orbit component would stop working, since the var we use to track moved calls would get out of wack with how many moved calls had actually happen

TODO:

- [x] WTF /datum/forced_movement exists and it's just my system but worse and bespoke. Integrate it, you'll need a component
- [x] Audit sleep() more closely. I feel like there's more of these (Most are gone now, only thing that comes to mind is disposal holders)
- [x] Everything that used the walk procs now has to deal with space movement. Do a pass on this.

## Why It's Good For The Game
Fixes meteors slogging the game down, kills walk(), makes spacedrift nice, and opens us up for some other nice optimizations and behavior improvements in future (Many of which I've already done because my brain is broken).
Oh and it gives atmos more cpu time. Fuck you.
## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: Space movement, assuming I've done my job correctly, should feel significantly smoother
fix: Fixed the immovable rod not properly carrying orbiters to new z levels 
code: Added a framework for different forms of looping movement
balance: Secborgs can no longer fly in space, as they have canonical wheels. The same likely applies to other spacefareing simple mobs, report anything that looks weird.
refactor: Refactored how meteors, some mobs, space drifting things, and a lot of general effects move, among other things. Please report any bugs or strange behavior
server: Hello admins. Spamming meteors will no longer lag things out quite so much. Still not amazing cause of all the explosions, but it won't kill things anymore. Blow some people up for me yeah?
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Adds a subsystem to handle automated directional movement, replaces all instances of walk_towards with it. Makes meteors and immovable rods not drift in space, and makes immovable rods more destructive. Note, I've opted not to use byond's method of moving towards something, which is effectively Move(src, get_step(src, get_dir(src, target))) as it's cringe and doesn't make a smooth line. I've replaced it with a autoupdating rise over run setup, read the code for more details

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
woop forgot the subsystem

LemonInTheDark committed:
Documentation, contributing.md entry, and some cleanup

LemonInTheDark committed:
Makes the moveloop datum more oop friendly, sets us up for a lot of conversions

LemonInTheDark committed:
Converts the curseblob and walk_away() to the subsystem

LemonInTheDark committed:
Changes the default for override from FALSE to TRUE

LemonInTheDark committed:
converts walk() over, still need to add a replacement proc for it, but we didn't actually have anything that used the raw proc

LemonInTheDark committed:
converts the rest of walk_to() over, nearing the end now

LemonInTheDark committed:
cleans up some errors

LemonInTheDark committed:
Fully documents everything, fills in some missing movement types, uses the power of oop to make things cleaner, and typepaths longer

LemonInTheDark committed:
Finishes the contributing.md stuff

LemonInTheDark committed:
Done

LemonInTheDark committed:
Fefaults -> Defaults, can you tell I wrote this at 1AM?

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
resolves bubblegum issues

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Roh's suggestions

Co-authored-by: Rohesie <rohesie@gmail.com>

LemonInTheDark committed:
Merge branch 'meteor-subsystem' of https://github.com/LemonInTheDark/tgstation into meteor-subsystem

LemonInTheDark committed:
Cleanup

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Hey lemon, did you know that Destroy() lives on datums? ahhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Converts over the discrepencies created in my absense

LemonInTheDark committed:
HAHA FUCK YOU I PAY MY DUES

LemonInTheDark committed:
Whoops lost some stuff in the merge

LemonInTheDark committed:
Converts the system from seconds to deciseconds to make dealing with the api more sane

LemonInTheDark committed:
Some stuff I missed

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Makes movement an inheritable subsystem type, splits the moveloop file into two, one for the subsystem, and one for the datums

LemonInTheDark committed:
Makes a subsystem that handles directing movers out to other subsystems. It's a bit bad right now, but it's a
good first step. I think I'll move the move loop datum to a lazy var on mobs instead of an assoc list, don't
like lists.

Also makes the movement procs global, I'll move em to the /movement subsystem at some point or something like
that

LemonInTheDark committed:
Converts the existing uses of the procs over to the new format

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Adds support for subsystem precedence, so a type of A can override type B.
General cleanup, still kinda in debug mode but it's getting better

LemonInTheDark committed:
I'll admit I'm not too familiar with this, but I think it will work

LemonInTheDark committed:
Adds starting logic so movement types "pausing" makes any sense
Redoes how waiting is handled to make it based on world.time directly. I don't remember why. I think it's better
this way.

Adds a drifting movement type, moves space drift over to it.
Needs severe work before it's ready, too much info stored and modified on the moving object, see comment
Starts work on making drifting smooth

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Moves almost all space drifting vars over to signals on the movement datum
Properly implements glide size stuff for both the subsystem and the loops. Space drift will be smoother now.
It's not perfect, but it'll work just fine for now

Adds a way to override a client'd mob's glide size mid move, uses it to make entering a spacedrift look right
Adds a way to delay a client move outside of just move_delay, meant to be used for long periods, and setup such
that it doesn't make inputs persist

Adds flags to movement loops, alongside MOVELOOP_OVERRIDE_CLIENT_CONTROL, which blocks client movements while
the loop is firing, and for it's visual delay after

This means you can't exit a space drift until you hit the actual wall. This feels a lot better

Some general logic stuff, move() will return true/false if it succeeded or failed
Adds a stop_loop() proc that's called when a move loop is no longer active

Suck my nuts

LemonInTheDark committed:
Moves precedence to the loop instead of the subsystem

LemonInTheDark committed:
Moves drifting into a component, this lets me explictly block input after the move loop ends, so people can't
move the moment they functionally move onto a new tile
This is a bit underdeveloped currently, but that's a problem for another day

Cleans up some uses of move procs, fixes runtimes in metoer and curseblob code
Adds signals for stopping/starting a move loop, sending one for destroy is redundant.
Moves existing event signals from the movable being acted on to the loop itself, makes more sense this way
Makes the move handler return the created loop up the chain so we can register to it
Fixes a logic error in loop contesting code that lead to loops never actually being removed from subsystems
because they didn't know they should be.
Properly changes lifetime from a time to stop, to functionally an amount of moves to complete before stopping
Adds some new signals for pre/post loop process. This is to better tie into components.

I decided I didn't like the idea of tying all functionality to the loops themselves
The loop decides functionally how to move, components or just tied in signals can decide when/when not to move
and can modify properties of the loop

Making a new loop for things like atmos drift, something I'm interested in tackling in the future, seemed silly

LemonInTheDark committed:
Moves movement procs directly to the subsystem for better namespacing or whatever

LemonInTheDark committed:
Moves movement packets onto /atom/movable, no longer need the debugging
I've decided to not just put their contents fully onto atom movable, since it makes debugging on live much
harder, can't sdql for them anymore.

Fixes a runtime in meteor code, properly this time
Fixes a logic error in stop_looping
Makes move manager NO_INIT, because well, it doesn't init

LemonInTheDark committed:
Commits human sin, makes Recover() work properly for movement subsystems

LemonInTheDark committed:
Fixes immovable rod orbits not always working, they were returning too early in moved and fucking up the var we use to track move count, and thus not sending a signal properly

LemonInTheDark committed:
Reworks the curseblob to use signals more, and to not use override

LemonInTheDark committed:
Missed this in the movement ss commit

LemonInTheDark committed:
Removes override, makes having a higher or equal precedence take its place

LemonInTheDark committed:
Updates documentation

LemonInTheDark committed:
Cleans up some unused defines

LemonInTheDark committed:
Nukes the unused flags option

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Whoops forgot to qdel check

LemonInTheDark committed:
Removes an unused var I had for client move prevention before I started using a component

LemonInTheDark committed:
Let's do this properly

LemonInTheDark committed:
Modernizes meteor code to better match how explosions actually work currently

LemonInTheDark committed:
Some more cleanup

LemonInTheDark committed:
Cleans up effect code a little bit
Nukes the effect system's sleep loop, we use movement loops instead
As a part of that, instead of 1 timer per effect spawned, we react to loop failure and make it 1 timer per
effect system
This should reduce the amoumt of slowdown we see after mass lighting break

It's not everything, we're still making a timer per spark effect, but it cuts things down significantly

LemonInTheDark committed:
Updates explosions to not sleep

TemporalOroboros commented:
Nice work. I'm pretty sure that spray bottles use a sleep loop if you want to take a look at them too.

LemonInTheDark committed:
Adds support for modifying a loops delay post process, makes extinguisher code suck less then it does currently, nukes some more sleeps and timer loops

LemonInTheDark committed:
Converts water tank resin over to move loops rather then sleeps, minor behavior change mind, the cooldown starts on fire rather then on land, but I think that makes more sense anyway

LemonInTheDark commented:
> 
> 
> Nice work. I'm pretty sure that spray bottles use a sleep loop if you want to take a look at them too.

I see what you mean yeah. This one's a bit of a headache, I'll give it a look

LemonInTheDark committed:
compile and runtime fix

CRITAWAKETS commented:
oh this is why servers start to drift into lag hell when space bases are made

LemonInTheDark commented:
> 
> 
> oh this is why servers start to drift into lag hell when space bases are made, thanks for fixing it

That was likely unrelated, since space drift is actually fine, if a bit crummy visually

LemonInTheDark committed:
Fixes some runtimes, cleans up some code, ensures feature parity when it comes to logging

LemonInTheDark committed:
Prevents resin foam from space drifting

LemonInTheDark committed:
Adds support for flags back into the system, I need it for reasons

LemonInTheDark committed:
Updates move_towards to fix some bugs and resolve some inconsistent behavior, implements a flag that makes a loop's first move start instantly

LemonInTheDark committed:
Fixes extinguishers not actually transfering any reagents

LemonInTheDark committed:
Converts sprays to the new system. This does actually minorly change behavior, in that I've changed the order of spray actions from step -> sleep -> wash to step -> wash -> sleep, but I'm not terribly torn up about it because frankly I think it feels better

LemonInTheDark committed:
Converts grav catapults over to the new system

LemonInTheDark committed:
Converts trays over to moveloops

LemonInTheDark committed:
Converts robot streaking to move loops, the other two coming soon

TemporalOroboros commented on code/modules/reagents/reagent_containers/spray.dm:
```diff
@@ -81,56 +81,13 @@
-			return
-
-	qdel(reagent_puff)
+	var/datum/move_loop/our_loop = SSmove_manager.move_towards(reagent_puff, target, wait_step, timeout = range * wait_step, flags = MOVEMENT_LOOP_START_FAST)
```
I love how this is still homing, even if it's only really noticeable if you target someone on meth or something.

TemporalOroboros replied:
Why is this a review, it should be a comment.

LemonInTheDark committed:
Compile you won't. Also fixes a behavior issue with oil streaks

LemonInTheDark committed:
Does directional step_to properly, cleans up the other two streaking types

LemonInTheDark commented:
> 
> 
> Nice work. I'm pretty sure that spray bottles use a sleep loop if you want to take a look at them too.

Done

LemonInTheDark committed:
Converts step_trigger over, not that it's actually used anywhere. Changes how stoping a move works, you need to explicitly qdel, other the step is just considered to be ignored. This will make life easier later

LemonInTheDark committed:
Adds a jps movement loop. It's a bit bloaty, id is stupid, but it'll work just fine

LemonInTheDark committed:
Makes the system support passing in a datum that's just used as extra context for the move. The hope is this makes signalizing things less of an absolute headache

LemonInTheDark committed:
Begins the conversion of ai movement datums to movement loops

LemonInTheDark committed:
These two are reasonably simple, only weird thing I'm doing is A: Not allowing target hotswapping, which I hope none is doing, and B: passing the controller into the move loop as extra context so things work properly

LemonInTheDark committed:
JPS is a bit more complex, partially because the old implementation was a bit weird. 2 major things. 1: I'm dropping what I think was a redundant behavior minimum distance check from the premove bit of logic, since I'm pretty sure it didn't do anything. 2, instead of just stoping the step in an error state like being pulled, we count it against our max move total

LemonInTheDark committed:
Audit

LemonInTheDark committed:
Moves most forced movement to the framework, adds some components to make things nicer

LemonInTheDark committed:
Implements a flag that makes the loop always operate, regardless of precedence and without impacting any other loops

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Moves movement subsystems into the right folder

LemonInTheDark committed:
Hey potato what if you had two procs that did the same thing and one called the other? Wow it's useless

LemonInTheDark committed:
Merges slipping and force movement

LemonInTheDark committed:
Converys conveyors over to the system. It's a bit fragile, but I think it's totally worth it to save the sleep loop

LemonInTheDark committed:
Precedence -> Priority, cleans up some logic errors, makes priority highest to lowest instead of lowest to highest, straight cleans some code up

LemonInTheDark committed:
Makes poly and bubbles ignore spacedrift, now that precedence actually functions properly. I'm likely missing cases of this, will deal with it later

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Depression, thy name is linter

LemonInTheDark committed:
Fixes linter, and hopefully fixes the runtimes in ci too

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Wew

LemonInTheDark committed:
Sets sprays and extinguishers back to legacy, since people do actually seem to have noticed

Kylerace reviewed (changes requested):
first round of reviews

LemonInTheDark commented on code/controllers/subsystem/movement/movement.dm:
```diff
@@ -0,0 +1,51 @@
+SUBSYSTEM_DEF(movement)
+	name = "Movement Loops"
+	flags = SS_BACKGROUND|SS_NO_INIT|SS_KEEP_TIMING
+	wait = 1 //Fire each tick
```
Right now my loops work off deciseconds.
If I'm honest with you, I don't really know how SS_TICKER works in regards to that. Could in theory use higher resolution though. 
I also really don't want the subsystems running first, things like spacedrift especially. May have some misconceptions here though.

Kylerace replied:
yeah that makes sense

LemonInTheDark commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -0,0 +1,641 @@
+	var/atom/movable/moving
+	///Defines how different move loops override each other. Lower numbers beat higher numbers
+	var/priority = MOVEMENT_DEFAULT_PRIORITY
+	///Bitfield of different things that effect how a loop operates
```
Tricked again

LemonInTheDark committed:
Spelling errors my beloved

Co-authored-by: Kylerace <kylerlumpkin1@gmail.com>

LemonInTheDark commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -0,0 +1,641 @@
+ * priority - Defines how different move loops override each other. Lower numbers beat higher numbers, equal defaults to what currently exists. Defaults to MOVEMENT_DEFAULT_PRIORITY
+ * flags - Set of bitflags that effect move loop behavior in some way. Check _DEFINES/movement.dm
+ *
+ * Returns TRUE if the loop sucessfully started, or FALSE if it failed
```
You mean the return values?

Kylerace replied:
yes

LemonInTheDark replied:
Feels like the wrong order to me, I think in terms of what's it do -> what does it want -> what does it give
But if you'd prefer it I'm alright doing that

LemonInTheDark commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -0,0 +1,641 @@
+	///The subsystem we're processing on
+	var/datum/controller/subsystem/movement/controller
+	///An extra reference we pass around because it's occasionally handy for signals
+	///This feels horrible, but constantly making components seems worse
```
I'm kinda adverse to giving exact examples, since comments don't compile. What in particular are you after? More detail on why you may want this?

Kylerace replied:
yes

LemonInTheDark replied:
Yeet

LemonInTheDark commented on code/controllers/subsystem/movement/movement.dm:
```diff
@@ -0,0 +1,51 @@
+		canonical_time = world.time
+		currentrun = processing.Copy()
+
+	var/list/running = currentrun //Cache for... you've heard this before
```
So true

LemonInTheDark committed:
More detail, moves return descriptions

LemonInTheDark committed:
Merge branch 'meteor-subsystem' of https://github.com/LemonInTheDark/tgstation into meteor-subsystem

JohnFulpWillard reviewed (approved):
[code_review.gif]

LemonInTheDark commented:
My botcode could beat up your botcode in a fight

LemonInTheDark committed:
Converts transit tubes to the system?

LemonInTheDark committed:
Adds the glide size modifier. Not honestly sure that this should be default, considering how crummy it makes things look for normal walking, but it's useful as hell here

LemonInTheDark committed:
Adds a force move in dir template, actual support for fast initial steps (wtf old me) and a helper proc for setting delay

LemonInTheDark committed:
Cleans up displosal code a bit, I thought about adding it to the system but it would functionally be just 'disposal loops'. Maybe I'll make a template subtype? not sure how I want to handle stuff like this

LemonInTheDark committed:
Cleans up mob movement a bit

LemonInTheDark committed:
Let's use the controller's visual delay

LemonInTheDark committed:
Makes the resin thrower nicer, cries

Kylerace commented on code/modules/recycling/conveyor.dm:
```diff
@@ -113,67 +157,63 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	conveying = TRUE
-
-	addtimer(CALLBACK(src, .proc/convey, affecting), 1)//Movement effect
+	moving.AddComponent(/datum/component/convey, movedir, 0.2 SECONDS)
```
i wonder if you can make this into just a dedicated movement packet instead of a component

LemonInTheDark replied:
It would show as an issue if someone somehow set things up to move/remove things from a conveyor really fast. But that's hard to do, and doesn't seem to be too bad? even in testing.

I'm trying to keep packets as pure as possible, rather then having them manage everything. This makes some things easier (space movement ending lag for instance) and I hope keeps things a bit cleaner

LemonInTheDark replied:
On the above note, this is why I haven't converted disposals despite them using sleep loops. They'd need their own unique move type just for them, still working out how I want to handle that

Kylerace commented on code/modules/meteors/meteors.dm:
```diff
@@ -2,7 +2,7 @@
 
 GLOBAL_VAR_INIT(meteor_wave_delay, 625) //minimum wait between waves in tenths of seconds
-//set to at least 100 unless you want evarr ruining every round
+//set to at least 100 unless you want evarr ruining every round (This spelling mistake is older then git, I'm scared to touch it)
```
>changes the git blame by adding something to the same line

LemonInTheDark replied:
Thank you for the catch, I'll move it to the line below.
I don't think the initial blame itself is that old, but may as well preserve

LemonInTheDark commented on code/datums/components/force_move.dm:
```diff
@@ -0,0 +1,38 @@
+/datum/component/force_move
+
+/datum/component/force_move/Initialize(atom/target, spin)
+	if(!target && ismob(parent))
```
I assume `!target || !ismob(parent)`, ye

LemonInTheDark commented on code/modules/recycling/conveyor.dm:
```diff
@@ -113,67 +157,63 @@ GLOBAL_LIST_EMPTY(conveyors_by_id)
-	var/list/items = locturf.contents - src
-	if(!LAZYLEN(items))//Dont do anything at all if theres nothing there but the conveyor
+/obj/machinery/conveyor/proc/start_conveying(atom/movable/moving)
+	if(!istype(moving) || iseffect(moving) || isdead(moving) || moving == src)
```
sure

LemonInTheDark replied:
Didn't want to do all of it, but did what I could

LemonInTheDark commented on code/__DEFINES/movement.dm:
```diff
@@ -14,3 +14,21 @@ GLOBAL_VAR_INIT(glide_size_multiplier, 1.0)
 #define DELAY_TO_GLIDE_SIZE(delay) (clamp(((32 / max((delay) / world.tick_lag, 1)) * GLOB.glide_size_multiplier), MIN_GLIDE_SIZE, MAX_GLIDE_SIZE))
+
+///Similar to DELAY_TO_GLIDE_SIZE, except without the clamping, and it supports piping in an unrelated scalar
+#define MOVEMENT_ADJUSTED_GLIDE_SIZE(delay, movement_disparity) (32 / ((delay) / world.tick_lag) * movement_disparity * GLOB.glide_size_multiplier)
```
in theory, yes. S hardcoded for now it seems. Could make it a define I guess? or just use it directly, same as tick_lag

LemonInTheDark commented on code/controllers/subsystem/movement/movement_types.dm:
```diff
@@ -0,0 +1,688 @@
+	return FALSE
+
+///Removes the atom from some movement subsystem. Defaults to SSmovement
+/datum/controller/subsystem/move_manager/proc/stop_looping(atom/movable/moving, datum/controller/subsystem/movement/subsystem = SSmovement)
```
I tried putting the starting procs in a separate file, and didn't like how disconnected things became
All but 1 of these are effectively helpers for the system below, all are "public"

stop_looping was mostly just cause it made sense given where all the rest were, but I'd really like to keep the movement type's definition near its helper proc, if only so people can see what the documentation is talking about directly. Plus it's better then needing to file hop to link a helper proc to what it really for real does.

Could be wrong about that, but I tried both ways and this seems more natural

LemonInTheDark committed:
Cleans up some comments, replaces an implicit world.icon_size with an explicit one, fixes up a typecheck

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
typecache instead of double istype. Can't do much about the !atom/movable, list would be too big I feel

LemonInTheDark committed:
hhh

LemonInTheDark committed:
bro wtf

LemonInTheDark commented:
Bruh

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Documents the why of SS_TICKER

LemonInTheDark committed:
Puts SSmovement on SS_TICKER. Lets us support tick steps

LemonInTheDark committed:
Cleans up the charge action. Makes it use moveloops

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Fixes CI? kinda worried that this just got dropped

LemonInTheDark committed:
Converts disposal pipes to move loops. They stutter a bit more then usual as of now, hoping that's a me thing, if it's not I'ma look at uping the priority of the base subsystem

LemonInTheDark committed:
Moves the move subsystems off background, puts some on ssticker

LemonInTheDark committed:
Prevents some things that shouldn't move in space from moving in space

LemonInTheDark committed:
Documents the general form and usage of the system

Kylerace commented on code/game/atoms_movable.dm:
```diff
@@ -960,14 +964,14 @@
-	SSspacedrift.processing[src] = src
+
+	set_glide_size(MOVEMENT_ADJUSTED_GLIDE_SIZE(inertia_move_delay, SSspacedrift.visual_delay))
+	AddComponent(/datum/component/drift, direction)
```
fear, have you profiled this?

LemonInTheDark replied:
I did some testruns and things didn't seem too off.
It's gonna be more expensive because we're running every tick rather then every 10, but it makes things look so much nicer. Don't think it's that bad.

I don't think the actual component add is too big a deal. It's more expensive, but it's a one time thing, and not too common of a one time thing.

Will get you better numbers on both the tick to tick operation and component stuff

LemonInTheDark replied:
Tested by dumping 1200 candy cigs out an airlock in cargo
Component add is cheap it seems
![image](https://user-images.githubusercontent.com/58055496/148873022-c5e7b8af-180f-4e0c-ab02-47be894dc031.png)
Self cost is high, but not horrid. Total's skewed cause I dumped em all out at once, so there's a lot of like (Increases with tile contents count) suffering
![image](https://user-images.githubusercontent.com/58055496/148872972-8ccabb08-d588-4184-96b3-23884c4bdf94.png)

LemonInTheDark replied:
Here was a test done by dumping 1200 items out an airlock with gas pressure
Done on both old and new code, new split up a bit which I expected would drop Move()'s total cost, but it didn't.

LemonInTheDark replied:
[old.csv](https://github.com/tgstation/tgstation/files/7845698/old.csv)
[new.csv](https://github.com/tgstation/tgstation/files/7845699/new.csv)

Kylerace reviewed (changes requested):
one more round

LemonInTheDark commented on code/datums/actions/mobs/charge.dm:
```diff
@@ -92,70 +118,59 @@
-					continue
-				T.attack_animal(charger)
+		for(var/obj/object in next_turf.contents)
+			if(!object.Adjacent(charger))
```
Very technically no. I'm tempted to leave answering this one to future scholars
![image](https://user-images.githubusercontent.com/58055496/148860430-4a26f41f-3da7-4c40-8450-44ccc8c5cf34.png)

LemonInTheDark replied:
Actually no that should be there, because the above check is more complex

LemonInTheDark commented on code/datums/ai/_ai_controller.dm:
```diff
@@ -261,3 +259,15 @@ multiple modular subtrees with behaviors
+	// right now I'm just taking the shortest minimum distance of our current behaviors, at some point in the future
+	// we should let whatever sets the current_movement_target also set the min distance and max path length
+	// (or at least cache it on the controller)
+	if(LAZYLEN(current_behaviors))
```
?

LemonInTheDark replied:
Oh right yeah uh, that can go

LemonInTheDark committed:
Virgin one vs chad once

Co-authored-by: Kylerace <kylerlumpkin1@gmail.com>

LemonInTheDark commented on code/controllers/subsystem/movement/ai_movement.dm:
```diff
@@ -1,20 +1,19 @@
+MOVEMENT_SUBSYSTEM_DEF(ai_movement)
 	name = "AI movement"
-	flags = SS_KEEP_TIMING|SS_BACKGROUND
+	flags = SS_KEEP_TIMING|SS_TICKER
```
Not having SS_BACKGROUND is more because I wasn't sure of the point of it, though I suppose that means I should just keep it.
Move looks need to be SS_TICKER because I do kinda want the behavior that provides. So A: cleanish tick based waits, and B: We do all our loop processing early, as if we were processing later. Saves some sanity. Though honestly maybe I don't want that. Huh. Lemme do this the hack way.

LemonInTheDark replied:
Changed my mind. See commit
https://github.com/tgstation/tgstation/pull/62567/commits/f6ddeeafb915c1e67fe2c14044a75579f6117209

LemonInTheDark committed:
Removes unneeded check

LemonInTheDark committed:
Moves appropriate movement subsystems into SS_BACKGROUND. Removes redundant SS_KEEP_TIMINGs

I do want the behavior of SS_TICKER, which at this point is tick based waits, and ignoring overtime when
calculating next fire.
Since honestly, these subsystems should ignore overtime in regards to next fire, the cost of moving A may be
nothing compared to the cost of moving B.

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark commented:
I've seen some buggy behavior like clipping through walls with move_towards. Need to fix that before a merge

LemonInTheDark committed:
Makes the MODULUS macro use floor. I knew our coders would never let me down, glad this exists, thanks ninja
Fixes teleporting caused by shitty round() behavior, adds a "you hit your target" case to homing loops

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into meteor-subsystem

LemonInTheDark committed:
Converts blood splatters to move loops, that'll do it

LemonInTheDark commented:
Fixed

Kylerace commented:
lets get it

LemonInTheDark commented:
It's finally done
